### ISSUE TITLE
activate Anaconda python3 environment does not work in .slurm file 

### ISSUE THREAD

**Issue by mkhoshle** [conda/conda#5071]
I have created an Anaconda python3 environment and I want to run my code using this version. But the problem is that when I run my script and do source activate inside my submit script, it does nothing. I am doing some parallel tasks and the problem is that my processes on the local node do not see any python 3 environment.

Here is what I have:
#!/usr/bin/env bash

source activate py35
my_file=PSA

SCHEDULER=`hostname`
echo SCHEDULER: $SCHEDULER
dask-scheduler --port=8786 &
sleep 5

hostnodes=`scontrol show hostnames $SLURM_NODELIST`
echo $hostnodes

for host in $hostnodes; do
    echo "Working on $host ...."
    ssh $host dask-worker --nprocs 24 --nthreads 1 $SCHEDULER:8786 &
    sleep 1
done

echo "====-get to work-===="

ssh $SCHEDULER
python $my_file.py $SCHEDULER:8786

Any idea of why this is happening?

**Comment by nehaljwani** [comment]
Did you try replacing...
```
source activate py35
```
...with:
```
source /absolute/path/to/directory/where/conda/is/installed/bin/activate py35
```

**Comment by mkhoshle** [comment]
Yes, I did. But still, it does not work.

**Comment by nehaljwani** [comment]
@mkhoshle Try replacing...
```
python $my_file.py $SCHEDULER:8786
```
...with:
```
/absolute/path/to/directory/where/conda/is/installed/envs/py35/bin/python $my_file.py $SCHEDULER:8786
```

**Comment by mkhoshle** [comment]
@nehaljwani   I tried this, still, it does not work. The problem is that when I print the version of the packages by client, scheduler and workers I get this:
```
{'client': {'packages': {'optional': [('numpy', '1.12.1'), ('pandas', '0.19.2')], 'required': [('dask', '0.14.1'), ('distributed', '1.16.1'), ('msgpack', '0.4.8'), ('cloudpickle', '0.2.2'), ('toolz', '0.8.2')]}, 'host': [('python', '3.5.3.final.0'), ('python-bits', 64), ('OS', 'Linux'), ('OS-release', '2.6.32-642.13.1.el6.x86_64'), ('machine', 'x86_64'), ('processor', 'x86_64'), ('byteorder', 'little'), ('LC_ALL', 'None'), ('LANG', 'en_US.UTF-8'), ('LOCALE', 'en_US.UTF-8')]}, 'scheduler': {'packages': {'optional': [['numpy', '1.12.1'], ['pandas', '0.19.2']], 'required': [['dask', '0.14.1'], ['distributed', '1.16.1'], ['msgpack', '0.4.8'], ['cloudpickle', '0.2.2'], ['toolz', '0.8.2']]}, 'host': [['python', '3.5.3.final.0'], ['python-bits', 64], ['OS', 'Linux'], ['OS-release', '2.6.32-642.13.1.el6.x86_64'], ['machine', 'x86_64'], ['processor', 'x86_64'], ['byteorder', 'little'], ['LC_ALL', 'None'], ['LANG', 'en_US.UTF-8'], ['LOCALE', 'en_US.UTF-8']]}, 'workers': {}}
```
```
I do not really understand why my worker processes started at each node do not show anything. When I run this with python 2, the workers show the version for all packages.
```

**Comment by kalefranz** [comment]
```
{
  'client':{
    'packages':{
      'optional':[
        ('numpy',
        '1.12.1'        ),
        ('pandas',
        '0.19.2'        )
      ],
      'required':[
        ('dask',
        '0.14.1'        ),
        ('distributed',
        '1.16.1'        ),
        ('msgpack',
        '0.4.8'        ),
        ('cloudpickle',
        '0.2.2'        ),
        ('toolz',
        '0.8.2'        )
      ]
    },
    'host':[
      ('python',
      '3.5.3.final.0'      ),
      ('python-bits',
      64      ),
      ('OS',
      'Linux'      ),
      ('OS-release',
      '2.6.32-642.13.1.el6.x86_64'      ),
      ('machine',
      'x86_64'      ),
      ('processor',
      'x86_64'      ),
      ('byteorder',
      'little'      ),
      ('LC_ALL',
      'None'      ),
      ('LANG',
      'en_US.UTF-8'      ),
      ('LOCALE',
      'en_US.UTF-8'      )
    ]
  },
  'scheduler':{
    'packages':{
      'optional':[
        [
          'numpy',
          '1.12.1'
        ],
        [
          'pandas',
          '0.19.2'
        ]
      ],
      'required':[
        [
          'dask',
          '0.14.1'
        ],
        [
          'distributed',
          '1.16.1'
        ],
        [
          'msgpack',
          '0.4.8'
        ],
        [
          'cloudpickle',
          '0.2.2'
        ],
        [
          'toolz',
          '0.8.2'
        ]
      ]
    },
    'host':[
      [
        'python',
        '3.5.3.final.0'
      ],
      [
        'python-bits',
        64
      ],
      [
        'OS',
        'Linux'
      ],
      [
        'OS-release',
        '2.6.32-642.13.1.el6.x86_64'
      ],
      [
        'machine',
        'x86_64'
      ],
      [
        'processor',
        'x86_64'
      ],
      [
        'byteorder',
        'little'
      ],
      [
        'LC_ALL',
        'None'
      ],
      [
        'LANG',
        'en_US.UTF-8'
      ],
      [
        'LOCALE',
        'en_US.UTF-8'
      ]
    ]
  },
  'workers':{

  }
}
```

**Comment by kalefranz** [comment]
Seems like a question @mrocklin might have a quick answer for?

**Comment by mrocklin** [comment]
I see that you've taken special care to start your client and scheduler in Python 3.  Perhaps you need to take the same care with your workers?  When you ssh into your different hosts and run `dask-worker` perhaps in that ssh session you are still using your system Python 2 installation.  When the Python 2 worker tries to connect up to the Python 3 scheduler it fails, which is why you don't see anything.

One solution to this would be to put the `source activate py35` line in your `.bash_profile` or `.bashrc` files.  This way all sessions, including ssh sessions, would run under python 3.5 by default.

**Comment by mkhoshle** [comment]
@mrocklin  I added `source activate py35` to both `.bashrc` and `.bash_profile` files. Still, I do not see anything for workers. I made some change to my submit script and I tried to activate the python3 environment on each host node when I ssh into them (and before starting dask-workers) and here is how it looks like:

```
export WKDIR=`pwd`

source /home/mkhoshle/miniconda2/envs/py35/bin/activate py35

my_file=PSA_Parallel
#my_file=test

export SCHEDULER=`hostname`
echo SCHEDULER: $SCHEDULER
dask-scheduler --port=8786 &
sleep 5

hostnodes=`scontrol show hostnames $SLURM_NODELIST`
echo $hostnodes

for host in $hostnodes; do
    echo "Working on $host ...."
    ssh $host $WKDIR/workers.sh $SCHEDULER &
    sleep 1
done

echo "====-get to work-===="

echo "Launching $my_file.py"
/home/mkhoshle/miniconda2/envs/py35/bin/python $my_file.py $SCHEDULER:8786
```

inside workers.sh I have 

```
#!/usr/bin/env bash

source /home/mkhoshle/miniconda2/envs/py35/bin/activate py35

echo "CHECKING dask-worker: $(which dask-worker)"
echo "CHECKING python: $(python --version 2>&1)"

echo "SCHEDULER: $1"
dask-worker --nprocs 24 --nthreads 1 ${1}:8786
```
and it shows the following in the output:

```
CHECKING dask-worker: /home/mkhoshle/miniconda2/envs/py35/bin/dask-worker
CHECKING python: Python 3.5.3 :: Continuum Analytics, Inc.
```

Also, when I do sys.path in my python script I get this:

```
['/oasis/scratch/comet/mkhoshle/temp_project/PSA-data2', '/opt/sdsc/lib',
 '/opt/sdsc/lib/python2.6/site-packages',
 '/home/mkhoshle/miniconda2/envs/py35/lib/python35.zip', 
'/home/mkhoshle/miniconda2/envs/py35/lib/python3.5', 
'/home/mkhoshle/miniconda2/envs/py35/lib/python3.5/plat-linux',
 '/home/mkhoshle/miniconda2/envs/py35/lib/python3.5/lib-dynload', 
'/home/mkhoshle/miniconda2/envs/py35/lib/python3.5/site-packages', 
'/home/mkhoshle/miniconda2/envs/py35/lib/python3.5/site-packages/setuptools-27.2.0-py3.5.egg']
```

**Comment by mkhoshle** [comment]
@mrocklin  I was getting this error when I was trying to send a copy of a large Numpy array to my other workers and it seems that the error was because of that problem. Because, my workers were not showing anything when I was printing Clinet.get_version(), I was thinking that they do not see python3 environment. I noticed that the problem was because of that array and when I split that array into smaller sizes I was able to run my code using python3 environment and I am able to see that my workers are showing the version of the all packages they see and they do see `'python','3.5.3.final.0'`. Also, there was no need for adding `source activate py35` to `.bashrc` file. Thanks all for all your help.
